ports/zephyr/modbluetooth_zephyr.c

```diff
@@ -312,24 +641,156 @@ int mp_bluetooth_gatts_write(uint16_t value_handle, const uint8_t *value, size_t
+        }
+    }
+
+    return bt_gatt_attr_read(conn, attr, buf, len, offset, entry->data, entry->data_len);
+}
+
+static ssize_t mp_bt_zephyr_gatts_attr_write(struct bt_conn *conn, const struct bt_gatt_attr *attr, const void *buf, uint16_t len, uint16_t offset, uint8_t flags) {
+    struct bt_conn_info info;
+    bt_conn_get_info(conn, &info);
+
+    DEBUG_printf("BLE attr write for handle %d\n", attr->handle);
+
+    // the characteristic handle is the value handle minus 1
+    uint16_t _handle = attr->handle - 1;
+
+    mp_bluetooth_gatts_db_entry_t *entry = mp_bluetooth_gatts_db_lookup(MP_STATE_PORT(bluetooth_zephyr_root_pointers)->gatts_db, _handle);
+    if (!entry) {
+        // it could be a descriptor instead
+        _handle = attr->handle;
+        entry = mp_bluetooth_gatts_db_lookup(MP_STATE_PORT(bluetooth_zephyr_root_pointers)->gatts_db, _handle);
+        if (!entry) {
+            return BT_GATT_ERR(BT_ATT_ERR_INVALID_HANDLE);
+        }
+    }
+
+    // Don't write anything if prepare flag is set
+    if (flags & BT_GATT_WRITE_FLAG_PREPARE) {
+        return 0;
     }
-    return mp_bluetooth_gatts_db_write(MP_STATE_PORT(bluetooth_zephyr_root_pointers)->gatts_db, value_handle, value, value_len);
+
+    if (offset > entry->data_alloc) {
+        return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+    }
+
+    if ((offset + len) > entry->data_alloc) {
+        return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+    }
+
+    if (entry->append) {
+        offset = entry->data_len;
+    }
+
+    // copy the data into the buffer in the gatts database
+    memcpy(&entry->data[offset], buf, len);
+    entry->data_len = offset + len;
+
+    mp_bluetooth_gatts_on_write(info.id, _handle);
+
+    return len;
+}
+
+static struct bt_gatt_attr *mp_bt_zephyr_find_attr_by_handle(uint16_t value_handle) {
+    for (int i = 0; i < MP_STATE_PORT(bluetooth_zephyr_root_pointers)->n_services; i++) {
+        int j = 0;
+        while (MP_STATE_PORT(bluetooth_zephyr_root_pointers)->services[i]->attrs[j].uuid != NULL) {
+            if (MP_STATE_PORT(bluetooth_zephyr_root_pointers)->services[i]->attrs[j].handle == value_handle) {
+                return &MP_STATE_PORT(bluetooth_zephyr_root_pointers)->services[i]->attrs[j];
+            }
+            j++;
+        }
+    }
+    return NULL;
 }
 
 int mp_bluetooth_gatts_notify_indicate(uint16_t conn_handle, uint16_t value_handle, int gatts_op, const uint8_t *value, size_t value_len) {
     if (!mp_bluetooth_is_active()) {
         return ERRNO_BLUETOOTH_NOT_ACTIVE;
     }
-    return MP_EOPNOTSUPP;
+
+    int err = MP_ENOENT;
+    mp_bt_zephyr_conn_t *connection = mp_bt_zephyr_find_connection(conn_handle);
+
+    if (connection) {
+        struct bt_gatt_attr *attr_val = mp_bt_zephyr_find_attr_by_handle(value_handle + 1);
+        mp_bluetooth_gatts_db_entry_t *entry = mp_bluetooth_gatts_db_lookup(MP_STATE_PORT(bluetooth_zephyr_root_pointers)->gatts_db, value_handle + 2);
```

Comment:
This `entry` variable is no longer needed.

In fact, I think the commit "ports/zephyr: Only send notifications/indications when subscribed." can be almost fully reverted, because you no longer need to create this additional database entry at `value_handle + 2`.  Is that correct?

Reply:
Yes, that's true, that `entry` variable is no longer needed. Will remove it.

But some of the changes in that commit are needed in my opinion, because in the case of the function `mp_bluetooth_gatts_write` you only want to send the notification/indication if the subscription is actually in place. While in the case of `mp_bluetooth_gatts_notify_indicate` we want to send the notification/indication regardless of the subscription state.

Reply:
> But some of the changes in that commit are needed in my opinion, because in the case of the function `mp_bluetooth_gatts_write` you only want to send the notification/indication if the subscription is actually in place

Yes, OK, let's keep it that additional functionality.